[WebGL] WebGL contexts drawn into canvas contexts are upside-down when premultipliedAlpha is false

djg opened the pull request:
#### ae8167890120f40300bca996c18227adc4ca1f9a
<pre>
[WebGL] WebGL contexts drawn into canvas contexts are upside-down when premultipliedAlpha is false
<a href="https://bugs.webkit.org/show_bug.cgi?id=259306">https://bugs.webkit.org/show_bug.cgi?id=259306</a>
rdar://112472666

Reviewed by Mike Wyrzykowski and Dean Jackson.

After the changes in 262518@main, the drawing buffer no longer needs to be
flipped before creating a native image to be used by drawImage.

GraphicsContextGL::paintRenderingResultsToPixelBuffer has been updated to take a
boolean specifying whether the image should be flipped when returning the
PixelBuffer. This is because the HTMLCanvasElement::getImageData() function,
used when using a canvas as a source to gl.texImage2D(), still requires the data
to be flipped.

Added new tests that exercise both code paths to catch any future regressions.

* LayoutTests/fast/canvas/webgl/canvas-drawImage-expected.html: Added.
* LayoutTests/fast/canvas/webgl/canvas-drawImage.html: Added.
* LayoutTests/fast/canvas/webgl/canvas-getImageData-expected.html: Added.
* LayoutTests/fast/canvas/webgl/canvas-getImageData.html: Added.
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getImageData):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::paintRenderingResultsToPixelBuffer):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::paintRenderingResultsToPixelBuffer):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::withDrawingBufferAsNativeImage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(paintRenderingResultsToPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::paintRenderingResultsToPixelBuffer):

Canonical link: <a href="https://commits.webkit.org/266225@main">https://commits.webkit.org/266225@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9cb122fc41e25dcc9b2db0cd7f7974928d82147e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13271 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13584 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13917 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15008 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12632 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13350 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16092 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13611 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15320 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13438 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14096 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11215 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15626 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11384 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11970 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19028 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12459 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12137 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15354 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12638 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10495 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11909 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11834 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3237 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16231 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12480 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
```diff
@@ -327,7 +327,7 @@ messages -> RemoteGraphicsContextGL NotRefCounted Stream {
     void GetInternalformativ(uint32_t target, uint32_t internalformat, uint32_t pname, size_t paramsSize) -> (IPC::ArrayReference<int32_t> params) Synchronous
     void SetDrawingBufferColorSpace(WebCore::DestinationColorSpace arg0)
-    void PaintRenderingResultsToPixelBuffer() -> (RefPtr<WebCore::PixelBuffer> returnValue) Synchronous
+    void PaintRenderingResultsToPixelBuffer(bool arg0) -> (RefPtr<WebCore::PixelBuffer> returnValue) Synchronous
```
Would propose `arg0` be named `flipY` but that is very minor

grorg replied:
See my comment above about enum class.

djg replied:
> Would propose `arg0` be named `flipY` but that is very minor

This code is generated by @kkinnunen-apple `generate-gpup-webgl`, which picks the name.

mwyrzykowski commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
```diff
@@ -1482,11 +1482,11 @@
         m_context->setDrawingBufferColorSpace(arg0);
     }
-    void paintRenderingResultsToPixelBuffer(CompletionHandler<void(RefPtr<WebCore::PixelBuffer>&&)>&& completionHandler)
+    void paintRenderingResultsToPixelBuffer(bool arg0, CompletionHandler<void(RefPtr<WebCore::PixelBuffer>&&)>&& completionHandler)
```
`arg0` means `flipY` right? Should it be named as such?

djg replied:
> `arg0` means `flipY` right? Should it be named as such?

Also generated by `generate-gpup-webgl`.

mwyrzykowski commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
```diff
@@ -3060,11 +3060,11 @@ void RemoteGraphicsContextGLProxy::setDrawingBufferColorSpace(const WebCore::Des
 }
 
-RefPtr<WebCore::PixelBuffer> RemoteGraphicsContextGLProxy::paintRenderingResultsToPixelBuffer()
+RefPtr<WebCore::PixelBuffer> RemoteGraphicsContextGLProxy::paintRenderingResultsToPixelBuffer(bool arg0)
```
similar to other comments

grorg commented on LayoutTests/fast/canvas/webgl/canvas-drawImage-expected.html:
```diff
@@ -0,0 +1,42 @@
+      {
+          const gl = wtu.create3DContext(canvas);
+          const program = wtu.setupProgram(gl,
+                                           [wtu.setupSimpleTextureVertexShader(gl), "quadFragmentShader"],
```
No need to change this, but you probably didn't need a texture sample. The fragment shader could have been a ternary that draws red when gl_FragCoord.y < 0 and green otherwise.

grorg replied:
And then the expected.html wouldn't have even needed to use WebGL.

grorg commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
```diff
@@ -440,7 +440,7 @@ class WebGLRenderingContextBase : public GraphicsContextGL::Client, public GPUBa
     void prepareForDisplayWithPaint() final;
     void paintRenderingResultsToCanvas() final;
-    RefPtr<PixelBuffer> paintRenderingResultsToPixelBuffer();
+    RefPtr<PixelBuffer> paintRenderingResultsToPixelBuffer(bool flipY);
```
Sorry to be a stickler, but we tend to use something like:

```
enum class FlipY : uint8_t {
  NotFlipped,
  Flipped
}
```

grorg replied:
So that the bool parameter makes more sense at the call sites.

djg replied:
> So that the bool parameter makes more sense at the call sites.

"and I would have gotten away with it too if it weren't for you meddling kids"

djg committed:
[WebGL] WebGL contexts drawn into canvas contexts are upside-down when premultipliedAlpha is false
https://bugs.webkit.org/show_bug.cgi?id=259306
rdar://112472666

Reviewed by Mike Wyrzykowski and Dean Jackson.

After the changes in 262518@main, the drawing buffer no longer needs to be
flipped before creating a native image to be used by drawImage.

GraphicsContextGL::paintRenderingResultsToPixelBuffer has been updated to take a
boolean specifying whether the image should be flipped when returning the
PixelBuffer. This is because the HTMLCanvasElement::getImageData() function,
used when using a canvas as a source to gl.texImage2D(), still requires the data
to be flipped.

Added new tests that exercise both code paths to catch any future regressions.

* LayoutTests/fast/canvas/webgl/canvas-drawImage-expected.html: Added.
* LayoutTests/fast/canvas/webgl/canvas-drawImage.html: Added.
* LayoutTests/fast/canvas/webgl/canvas-getImageData-expected.html: Added.
* LayoutTests/fast/canvas/webgl/canvas-getImageData.html: Added.
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getImageData):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::paintRenderingResultsToPixelBuffer):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::paintRenderingResultsToPixelBuffer):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::withDrawingBufferAsNativeImage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(paintRenderingResultsToPixelBuffer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::paintRenderingResultsToPixelBuffer):

Canonical link: https://commits.webkit.org/266225@main
